scripts/dasharo-deploy

```diff
@@ -96,32 +197,194 @@ ask_for_version() {
         ;;
     esac
   done
+
+  return 0
+}
+
+choose_version(){
+# This function is used for determining Dasharo firmware update version and is
+# being used during updates only. We do not ask user to choose firmware update
+# versions compared to installation workflow (check ask_for_version function),
+# instead we have some priorities:
+# 1) Check if Dasharo Heads Firmware available, use it if yes;
+# 2) Check if Dasharo EDK2 Firmware available, use it if yes;
+# 3) Use Dasharo Community Firmware;
+#
+# Capsules have higher priority over simple binaries.
+#
+# TODO: Currently we do not have clear and concise update mechanisms (e.g. what
+# and when a specific firmware version can be used, how to handle revisions of
+# firmware).
+
+  if [ "$HAVE_HEADS_FW" == "true" ]; then
+    if check_for_firmware_access heads; then
+      CAN_SWITCH_TO_HEADS="true"
+      FIRMWARE_VERSION="heads"
+
+      return 0
+    else
+      print_firm_access_warning heads
+    fi
+  fi
+
+  if [ -n "$DASHARO_REL_VER_DPP_CAP" ]; then
```

Comment:
What if we are updating Dasharo from the version which does not support Capsule Update? DTS will automatically choose to use capsules and will fail every update, because it does not have any mechanisms to check, whether current firmware support capsule update.

From what I can see, the driver responsible for the update does not have functionality to provide this info before the update. So a variable should be added for each platform that support capsules to signal, form which version of firmware capsule update is supported.

This is a TODO note for me.

Reply:
in board config check if version is less than x and if it is then unset `DASHARO_REL_VER_DPP_CAP`?

Reply:
Yep, smth like it would be ok.

Reply:
Added here 9aa00a04166f712f638acc875e4004c28ece5c32.

Reply:
Wouldn't unsetting all [those variables](https://github.com/Dasharo/dts-scripts/blob/6cb56fe167dbe4a836069c550ebc5d03ba893a1f/include/dts-functions.sh#L672) work as well? It would be easier to keep track if it's only checked/set in one place e.g.

```diff
-  # And for capsules as well:
-  [ -z "$BIOS_HASH_LINK_COMM_CAP" ] && BIOS_HASH_LINK_COMM_CAP="${BIOS_LINK_COMM_CAP}.sha256"
-  [ -z "$BIOS_SIGN_LINK_COMM_CAP" ] && BIOS_SIGN_LINK_COMM_CAP="${BIOS_HASH_LINK_COMM_CAP}.sig"
-  [ -z "$BIOS_HASH_LINK_DPP_CAP" ] && BIOS_HASH_LINK_DPP_CAP="${BIOS_LINK_DPP_CAP}.sha256"
-  [ -z "$BIOS_SIGN_LINK_DPP_CAP" ] && BIOS_SIGN_LINK_DPP_CAP="${BIOS_HASH_LINK_DPP_CAP}.sig"
-  [ -z "$EC_HASH_LINK_COMM_CAP" ] && EC_HASH_LINK_COMM_CAP="${EC_LINK_COMM_CAP}.sha256"
-  [ -z "$EC_SIGN_LINK_COMM_CAP" ] && EC_SIGN_LINK_COMM_CAP="${EC_HASH_LINK_COMM_CAP}.sig"
-  [ -z "$EC_HASH_LINK_DPP_CAP" ] && EC_HASH_LINK_DPP_CAP="${EC_LINK_DPP_CAP}.sha256"
-  [ -z "$EC_SIGN_LINK_DPP_CAP" ] && EC_SIGN_LINK_DPP_CAP="${EC_HASH_LINK_DPP_CAP}.sig"
+  if compare_versions "$DASHARO_VERSION" "$DASHARO_SUPPORT_CAP_FROM" ; then
+    # And for capsules as well:
+    [ -z "$BIOS_HASH_LINK_COMM_CAP" ] && BIOS_HASH_LINK_COMM_CAP="${BIOS_LINK_COMM_CAP}.sha256"
+    [ -z "$BIOS_SIGN_LINK_COMM_CAP" ] && BIOS_SIGN_LINK_COMM_CAP="${BIOS_HASH_LINK_COMM_CAP}.sig"
+    [ -z "$BIOS_HASH_LINK_DPP_CAP" ] && BIOS_HASH_LINK_DPP_CAP="${BIOS_LINK_DPP_CAP}.sha256"
+    [ -z "$BIOS_SIGN_LINK_DPP_CAP" ] && BIOS_SIGN_LINK_DPP_CAP="${BIOS_HASH_LINK_DPP_CAP}.sig"
+    [ -z "$EC_HASH_LINK_COMM_CAP" ] && EC_HASH_LINK_COMM_CAP="${EC_LINK_COMM_CAP}.sha256"
+    [ -z "$EC_SIGN_LINK_COMM_CAP" ] && EC_SIGN_LINK_COMM_CAP="${EC_HASH_LINK_COMM_CAP}.sig"
+    [ -z "$EC_HASH_LINK_DPP_CAP" ] && EC_HASH_LINK_DPP_CAP="${EC_LINK_DPP_CAP}.sha256"
+    [ -z "$EC_SIGN_LINK_DPP_CAP" ] && EC_SIGN_LINK_DPP_CAP="${EC_HASH_LINK_DPP_CAP}.sig"
+  else
+    unset "$BIOS_HASH_LINK_COMM_CAP"
+    unset "$BIOS_SIGN_LINK_COMM_CAP"
+    unset "$BIOS_HASH_LINK_DPP_CAP"
+    unset "$BIOS_SIGN_LINK_DPP_CAP"
+    unset "$EC_HASH_LINK_COMM_CAP"
+    unset "$EC_SIGN_LINK_COMM_CAP"
+    unset "$EC_HASH_LINK_DPP_CAP"
+    unset "$EC_SIGN_LINK_DPP_CAP"
+  fi
```

also if you are using `compare_versions` then make sure `DASHARO_SUPPORT_CAP_FROM` always has correct version i.e. (<number>.<number>.<number>[rc[.]<number>] otherwise it'll error_exit https://github.com/Dasharo/dts-scripts/blob/9aa00a04166f712f638acc875e4004c28ece5c32/include/dts-functions.sh#L737

Reply:
> Wouldn't unsetting all [those variables](https://github.com/Dasharo/dts-scripts/blob/6cb56fe167dbe4a836069c550ebc5d03ba893a1f/include/dts-functions.sh#L672) work as well?

No, it will not, because `DASHARO_REL_VER*_CAP` variables are used for selecting update type.

> It would be easier to keep track if it's only checked/set in one place e.g.

This is subjective, I prefer to use `board_config` to set all needed configs, and let other functions to decide what to do with it. If we are using `choose_version` to choose update type, why not to check there?

> also if you are using compare_versions then make sure DASHARO_SUPPORT_CAP_FROM always has correct version

I am sure that I am using correct version everywhere.